### PR TITLE
Fix text cut off when halo text width is applied

### PR DESCRIPTION
**Pull request by Chaoba**
`<changelog>Fix the issue that top of the text is cut off when halo text width is applied.</changelog>`

### CODE REVIEW

File: MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/text/LocalGlyphRasterizer.java [hunk 1]
```diff
@@ -31,7 +31,7 @@ of the bitmap (y: 20) with some buffer around the edge
 
     paint = new Paint();
     paint.setAntiAlias(true);
-    paint.setTextSize(24);
+    paint.setTextSize(22);
```
Comment (tsuz) [inline MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/text/LocalGlyphRasterizer.java:34]: Will this cause an issue if a developer sets a larger font size via the app or via studio?

File: MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/text/LocalGlyphRasterizer.java [hunk 2]
```diff
@@ -31,7 +31,7 @@ of the bitmap (y: 20) with some buffer around the edge
 
     paint = new Paint();
     paint.setAntiAlias(true);
-    paint.setTextSize(24);
```
Comment (alexshalamov) [inline MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/text/LocalGlyphRasterizer.java:34]: @Chaoba All locally and server generated fonts are rendered at 24 points. Maybe you need to increase bitmap size. Issue needs bit more investigation, 24pt glyph should fit into 35x35 bitmap.

File: MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/text/LocalGlyphRasterizer.java (hunk 1 above)
Comment (Chaoba) [reply MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/text/LocalGlyphRasterizer.java:34]: What's the text size do you set in your style? Can you set it to a very large one to help me verify it?

Comment (alexshalamov) [reply MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/text/LocalGlyphRasterizer.java:34]: @tsuz @Chaoba `24pt` is a base font size across whole mapbox ecosystem, server and client side. This value should not be changed.

We use SDF rasterezed glyphs, they can be up/downscaled without much artifacts. I don't see how setting very large text size in studio would help here.

I think the issue is that for some reason, Android renders CJK glyph too close to a bitmap border (we expect to have have 3px padding around rendered glyph)
35x35px buffer comes from 24px + 3px border * 2 + 1px glyph padding * 2 + safe margin.

@Chaoba could you increase buffer size and see if truncation is fixed? You can also save bitmap locally and see if there is a truncation in SDF glyph.

**Comment by Chaoba** [comment]
Save some Chinese bitmap and we can find all characters are on the left top
image

And for `Noto Sans CJK JP Bold`, the characters are on left top too and some are cut off
image

We can move characters a litter right and bottom to fix this issue. @alexshalamov What do you think?

**Comment by Chaoba** [comment]
Characters will be in the center after modification. 
image

File: MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/text/LocalGlyphRasterizer.java (hunk 2 above)
Comment (Chaoba) [reply MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/text/LocalGlyphRasterizer.java]: Increate bitmap size will twist characters.

**Comment by chloekraw** [comment]
If this is possible to resolve by the same time as https://github.com/mapbox/mapbox-gl-native-android/issues/377 and we feel good about the resolution, let's cherrypick this into v9.2.1 as well.

**Comment by Chaoba** [comment]
The current solution will cause irregular while text contains numbers.
image

**Comment by alexshalamov** [comment]
> The current solution will cause irregular while text contains numbers.

@Chaoba that is because a glyph was moved within a bitmap and is out-of-sync with glyph metrics, therefore, CJK glyph baseline is 'lower' than latin character's baseline. Do you have a branch with a test that I can try?

**Comment by Chaoba** [comment]
Thanks to @alexshalamov, it looks good now.
image

**Comment by alexshalamov** [comment]
@1ec5 FYI, might be applicable to iOS / darwin code as well.

**Comment by chloekraw** [comment]
> maybe worth adding unit tests to catch regressions

Yes please, @Chaoba could you add a test?

**Comment by Chaoba** [comment]
It's not easy to verify this part by a test case. @tobrun Do you have any suggestions on it?

**Comment by alexshalamov** [comment]
> It's not easy to verify this part by a test case.

@tobrun Are there any examples of 'pixelmatch' tests that verify rendering result?

**Comment by tobrun** [comment]
> @tobrun Are there any examples of 'pixelmatch' tests that verify rendering result?

I had a small POC render test in https://github.com/mapbox/mapbox-gl-native/pull/14687 but that became obsolete with the platform specific c++ test-runner. This PR however highlights that we need to support platform specific pixelmatch testing. I will put some notes down into a separate issue for an Android instrumented unit test integration 

cc @julianrex

**Pull request merged**